storage/pin/pin_test.go

```diff
@@ -0,0 +1,328 @@
+func TestPinRawUploadEncrypted(t *testing.T) {
+
+	p, f := getPinApiAndFileStore(t)
+	data := testutil.RandomBytes(1, 10000)
+	hash := uploadFile(t, f, data, true)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 5, true)
+
+}
+
+// Pin a simple collection and unpin it multiple times
+func TestPinCollectionUpload(t *testing.T) {
+
+	p, f := getPinApiAndFileStore(t)
+	hash := uploadCollection(t, p, f, false)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 1, false)
+
+}
+
+// Pin a encrypted simple collection and unpin it multiple times
+func TestPinCollectionUploadEncrypted(t *testing.T) {
+
+	p, f := getPinApiAndFileStore(t)
+	hash := uploadCollection(t, p, f, true)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 5, false)
+
+}
+
+// Test if the uploaded collection shows up in the ListPinFiles command
+func TestListPinInfo(t *testing.T) {
+
+	p, f := getPinApiAndFileStore(t)
+	hash := uploadCollection(t, p, f, false)
+
+	// Pin the hash
+	err := p.PinFiles(hex.EncodeToString(hash), false, "")
+	if err != nil {
+		t.Fatalf("Could not pin " + err.Error())
+	}
+
+	// Gte the list of file pinned
+	pinInfo := p.ListPinFiles()
+
+	// Check if the uploaded collection is in the list files
+	if _, ok := pinInfo[hex.EncodeToString(hash)]; !ok {
+		t.Fatalf("uploaded collection not pinned")
+	}
+
+}
+
+func getPinApiAndFileStore(t *testing.T) (*PinAPI, *storage.FileStore) {
+
+	t.Helper()
+
+	swarmDir, err := ioutil.TempDir("", "swarm-storage-test")
+	if err != nil {
+		t.Fatalf("could not create temp dir. Error: %s", err.Error())
+	}
+
+	localStore, err := localstore.New(swarmDir, make([]byte, 32), nil)
+	if err != nil {
+		os.RemoveAll(swarmDir)
+		t.Fatalf("could not create localstore. Error: %s", err.Error())
+	}
+
+	tags := chunk.NewTags()
+	fileStore := storage.NewFileStore(localStore, storage.NewFileStoreParams(), tags)
+
+	// Swarm feeds test setup
+	feedsDir, err := ioutil.TempDir("", "swarm-feeds-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	feeds, err := feed.NewTestHandler(feedsDir, &feed.HandlerParams{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	swarmApi := api.NewAPI(fileStore, nil, feeds.Handler, nil, tags)
+
+	pinAPI := NewPinApi(localStore, nil, tags, swarmApi)
+
+	return pinAPI, fileStore
+
+}
+
+func uploadFile(t *testing.T, f *storage.FileStore, data []byte, toEncrypt bool) storage.Address {
+
+	t.Helper()
+
+	size := int64(len(data))
+	ctx := context.TODO()
+	addr, wait, err := f.Store(ctx, bytes.NewReader(data), size, toEncrypt)
+	if err != nil {
+		t.Fatalf("could not store file. Error: %s", err.Error())
+	}
+
+	err = wait(ctx)
+	if err != nil {
+		t.Fatalf("Store wait error: %v", err.Error())
+	}
+
+	return addr
+}
+
+type fileInfo struct {
+	fileName string
+	fileHash storage.Address
+}
+
+func uploadCollection(t *testing.T, p *PinAPI, f *storage.FileStore, toEncrypt bool) storage.Address {
```

Comment:
why go through all this hassle instead of just creating a temporary directory and a few files with some example content?
there's plenty of examples in `cmd/swarm`

Reply:
This is the same as "swarm up".. i don't see any hassle in this.  Also if you upload a dir it is difficult to know the manifest tree structure and hashes .. which is used to test the result..

This way is simple and clean.